modules/bezug_e3dc/e3dc.py

```diff
@@ -1,90 +1,37 @@
 #!/usr/bin/python
-import sys
-# import os
-# import time
-# import getopt
-# import socket
-# import ConfigParser
-import struct
-# import binascii
-from pymodbus.client.sync import ModbusTcpClient
+from pymodbus.constants import Endian
+import logging
+import math
+from typing import List
 
-ipaddress = str(sys.argv[1])
+from helpermodules.cli import run_using_positional_cli_args
+from helpermodules.log import setup_logging_stdout
+from modules.common.component_state import CounterState
+from modules.common.modbus import ModbusClient, ModbusDataType
+from modules.common.store import get_counter_value_store
 
-client = ModbusTcpClient(ipaddress, port=502)
+log = logging.getLogger("E3DC EVU")
 
-# evu punkt
-resp= client.read_holding_registers(40073,2,unit=1)
-value1 = resp.registers[0]
-value2 = resp.registers[1]
-all = format(value2, '04x') + format(value1, '04x')
-final = int(struct.unpack('>i', all.decode('hex'))[0])
-f = open('/var/www/html/openWB/ramdisk/wattbezug', 'w')
-f.write(str(final))
-f.close()
+def update(ipaddress: str):
+    log.debug("Beginning update")
+    client = ModbusClient(ipaddress, port=502)
+#40074 EVU Punkt negativ -> Einspeisung in Watt
+    power_all = client.read_holding_registers(40073, ModbusDataType.INT_32,wordorder=Endian.Little, unit=1)
+#40130 Phasenleistung in Watt
+    powers = [val for val in client.read_holding_registers(40129, [ModbusDataType.INT_16] * 3,unit=1)]
+    currents = [0]*3
+    voltages = [230]*3
+    for i in range(0, 3):
+        if powers[i] > 0:
+            currents[i]=powers[i]/voltages[i]
```

Comment:
Ich vermute es sollte `if powers[i] <> 0` heißen und nicht `if powers[i] > 0`?

Nur aus Spaß, alternativer Vorschlag:
```python
currents = [0 if p == 0 else p/230 for p in powers]
```

Reply:
Nee ,
Hier geht es darum das Ampere pro Phase nur dann gerechnet werden soll, wenn eine Bezug vorliegt. Bei einer Einspeisung (powers [L1,L2,l3] ist negativ)  soll keine Ampere für L1, L2, L3  gerechnet werden. Anderenfalls könnte die Schieflasterkennung fälschlicherweise anschlagen.

Reply:
Wo ist der Unterschied, ob die Schieflast zwischen zwei positiven oder einem positiven und einem negativen Strom vorliegt? Beides ist nicht erlaubt.

Reply:
Bisher war es zumindest um e3dc Modul anders, aber wir können es natürlich  ändern:
Okay nehmen wir an wir haben 
auf l1 690 Watt Bezug
auf l2 690 Watt Einspeisung
auf L3 690 Watt Bezug
Dann würde wir nach der Vorgaben 
a) auf L1 -> 3A und auf L2 -> **-3A** und auf L3 3A rechnen  ? (Schieflast nehme ich an 6 A)
b) auf L1 -> 3A und auf L2 -> 3A und auf L3 3A rechnen  ? (Schieflast nehme ich an 0A)
c) auf L1 -> 3A und auf L2 -> 0A  und auf L3 3A rechnen  ? (so passiert es heute) (Schieflast nehme ich an 3 A)
Welche Option ist richtig ?

Reply:
Option a.

Der Grund der Vermeidung von Schieflast liegt darin, dass es Komponenten "da draußen" (Kraftwerksgenerator) beschädigen könnte, wenn nicht auf allen 3 Phasen in etwa das Gleiche passiert. 3A und -3A sind nicht das Gleiche. Das ist ein ziemlich krasser Unterschied. Warum sollte es einen Unterschied machen ob du [8A, 2A, 2A] oder [7A, 1A, 1A] oder [6A, 0A, 0A] oder [5A, -1A, -1A] etc. hast?

Gut, ich bin kein Elektroingineur, man möge mich korrigieren, wenn ich da was völlig falsch verstanden habe.

Reply:
Passt für mich ich bau Option a ein und teste das morgen nochmal mit hoffentlich ein wenig Sonnenschein. Ich melde mich hier nochmal mit dem Ergebnis vom Test...

Reply:
Habe es getest, bei Einspeisung wird nun die Ampere Zahl negativ gerechnet. Aus meiner Sicht kann man das nun übernehmen. @yankee42 oder @benderl Irgendwelche Vorbehalte ?